dbms/src/Storages/MySQL/MySQLSettings.cpp

```diff
@@ -0,0 +1,137 @@
+#include <Storages/MySQL/MySQLSettings.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Common/StringUtils/StringUtils.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTLiteral.h>
+#include <Interpreters/evaluateConstantExpression.h>
+#include <Parsers/ExpressionElementParsers.h>
+#include <Parsers/parseQuery.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+static const String VARIABLE_PREFIX = "mysql_variable_";
+
+static void getFieldWithLiteralQueryString(Field & field, const String & literal_query_string)
+{
+    ParserLiteral literal_p;
+    std::string error_message;
+    auto literal_query_string_data = literal_query_string.data();
+    ASTPtr literal_node = tryParseQuery(literal_p, literal_query_string_data, literal_query_string.data() + literal_query_string.size(),
+                                        error_message, false, "", false, 0);
+
+    field = literal_node ? static_cast<ASTLiteral *>(literal_node.get())->value : Field(literal_query_string);
+}
+
+MySQLSettings::MySQLSettings(const ASTStorage & storage_def, const Context & context, const Poco::Util::AbstractConfiguration & config,
+                             const String & default_database, const String & default_table)
+{
+    if (storage_def.settings)
+        evaluateSetQuery(storage_def.settings);
+
+    if (!storage_def.engine->arguments->children.empty())
+        evaluateEngineArguments(storage_def.engine->arguments->children, context);
+
+    if (!remote_database.changed || remote_table_name.changed)
+        fillDefaultDatabaseAndTableName(default_database, default_table);
+
+    const auto & set_query = convertConfigToSetQuery(config, "mysql");
+    evaluateSetQuery(static_cast<const ASTSetQuery *>(set_query.get()));
+}
+
+void MySQLSettings::evaluateSetQuery(const ASTSetQuery * set_query)
+{
+    for (const ASTSetQuery::Change & setting : set_query->changes)
+    {
+#define MATCH_MySQL_SETTING(TYPE, NAME, DEFAULT, DESCRIPTION) \
```

Comment:
I suppose the class `MySQLSettings` should be derived from the template class `SettingsCollection` - it allows using here normal loops instead of these tricks with macros.